pkg/apis/eventing/v1alpha1/subscription_validation.go

```diff
@@ -66,15 +67,34 @@ func (ss *SubscriptionSpec) Validate(ctx context.Context) *apis.FieldError {
 }
 
 func isSubscriberSpecNilOrEmpty(s *SubscriberSpec) bool {
-	return s == nil || equality.Semantic.DeepEqual(s, &SubscriberSpec{}) ||
-		(equality.Semantic.DeepEqual(s.Ref, &corev1.ObjectReference{}) && s.DNSName == nil)
-
+	if s == nil || equality.Semantic.DeepEqual(s, &SubscriberSpec{}) {
+		return true
+	}
+	if equality.Semantic.DeepEqual(s.Ref, &corev1.ObjectReference{}) &&
+		s.DeprecatedDNSName == nil &&
+		s.URI == nil {
+		return true
+	}
+	return false
 }
 
 func isValidSubscriberSpec(s SubscriberSpec) *apis.FieldError {
 	var errs *apis.FieldError
-	if s.DNSName != nil && *s.DNSName != "" && s.Ref != nil && !equality.Semantic.DeepEqual(s.Ref, &corev1.ObjectReference{}) {
-		errs = errs.Also(apis.ErrMultipleOneOf("ref", "dnsName"))
+
+	fieldsSet := make([]string, 0, 0)
+	if s.Ref != nil && !equality.Semantic.DeepEqual(s.Ref, &corev1.ObjectReference{}) {
+		fieldsSet = append(fieldsSet, "ref")
+	}
+	if s.DeprecatedDNSName != nil && *s.DeprecatedDNSName != "" {
+		fieldsSet = append(fieldsSet, "dnsName")
```

Comment:
should it populate the `uri` ?

Reply:
My hesitation for doing that is code like our [Trigger controller](https://github.com/knative/eventing/blob/e245a22b50d40d9d12d8ad5f6ca93d9d0781f7f3/pkg/reconciler/v1alpha1/trigger/trigger.go#L563), which will compare the `spec`, if it differs, it will delete the `Subscription` and create a new one. Our Trigger controller wouldn't hit that problem because it uses `ref`, not `dnsName`, but I could see a similar controller hitting the problem.

What do you think? Is it worth the potential trouble to move things for the user?

Reply:
ok, fair point w/ the Trigger controller.

So, no need to "adjust" it